app/utils/url.ts

```
@@ -0,0 +1,82 @@
+export const parseQueryString = <T = string>(queryString: string): ParsedQuery<T> =>
+  typeof queryString === 'string' && queryString.length > 0
+    ? Array.from(
+        new URLSearchParams(queryString.startsWith('?') ? queryString.substring(1) : queryString).entries(),
+      ).reduce(
+        (obj, [key, val]) => {
+          const value = val || null; // value should fallback to `null` instead of empty string
+          // if key already exists, convert to array and append value:
+          const valueInParsedQueryObject = obj[key];
+          if (valueInParsedQueryObject)
+            return {
+              ...obj,
```

Biome (1.9.4)

[error] Avoid the use of spread (`...`) syntax on accumulators.

Spread syntax should be avoided on accumulators (like those in `.reduce`) because it causes a time complexity of `O(n^2)`.
Consider methods such as .splice or .push instead.

(lint/performance/noAccumulatingSpread)